pkg/zfs/plugin/zfs.go

```diff
@@ -0,0 +1,201 @@
+		p.incremental = true
+	}
+
+	conf, err := rest.InClusterConfig()
+	if err != nil {
+		p.Log.Errorf("Failed to get cluster config : %s", err.Error())
+		return errors.New("error fetching cluster config")
+	}
+
+	clientset, err := kubernetes.NewForConfig(conf)
+	if err != nil {
+		p.Log.Errorf("Error creating clientset : %s", err.Error())
+		return errors.New("error creating k8s client")
+	}
+
+	if err := velero.InitializeClientSet(conf); err != nil {
+		return errors.Wrapf(err, "failed to initialize velero clientSet")
+	}
+
+	p.K8sClient = clientset
+
+	p.cl = &cloud.Conn{Log: p.Log}
+	return p.cl.Init(config)
+}
+
+// CreateVolumeFromSnapshot creates a new volume from the specified snapshot
+func (p *Plugin) CreateVolumeFromSnapshot(snapshotID, volumeType, volumeAZ string, iops *int64) (string, error) {
+	p.Log.Infof("zfs: CreateVolumeFromSnapshot called snap %s", snapshotID)
+
+	volumeID, err := p.doRestore(snapshotID)
+
+	if err != nil {
+		p.Log.Errorf("zfs: error CreateVolumeFromSnapshot returning snap %s err %v", snapshotID, err)
+		return "", err
+	}
+
+	p.Log.Infof("zfs: CreateVolumeFromSnapshot returning snap %s vol %s", snapshotID, volumeID)
+	return volumeID, nil
+}
+
+// GetVolumeInfo returns the type and IOPS (if using provisioned IOPS) for
+// the specified volume in the given availability zone.
+func (p *Plugin) GetVolumeInfo(volumeID, volumeAZ string) (string, *int64, error) {
+	p.Log.Infof("zfs: GetVolumeInfo called", volumeID, volumeAZ)
+	return "zfs-localpv", nil, nil
+}
+
+// IsVolumeReady Check if the volume is ready.
+func (p *Plugin) IsVolumeReady(volumeID, volumeAZ string) (ready bool, err error) {
+	p.Log.Infof("zfs: IsVolumeReady called", volumeID, volumeAZ)
+
+	return p.isVolumeReady(volumeID)
+}
+
+// CreateSnapshot creates a snapshot of the specified volume, and applies any provided
+// set of tags to the snapshot.
+func (p *Plugin) CreateSnapshot(volumeID, volumeAZ string, tags map[string]string) (string, error) {
+	p.Log.Infof("zfs: CreateSnapshot called", volumeID, volumeAZ, tags)
+
+	bkpname, ok := tags[VeleroBkpKey]
+	if !ok {
+		return "", errors.New("zfs: error get backup name")
+	}
+
+	schdname, ok := tags[VeleroSchdKey]
+
+	snapshotID, err := p.doBackup(volumeID, bkpname, schdname)
+
+	if err != nil {
+		p.Log.Errorf("zfs: error createBackup %s@%s failed %v", volumeID, bkpname, err)
+		return "", err
+	}
+
+	p.Log.Infof("zfs: CreateSnapshot returning %s", snapshotID)
+	return snapshotID, nil
+}
+
+// DeleteSnapshot deletes the specified volume snapshot.
+func (p *Plugin) DeleteSnapshot(snapshotID string) error {
+	p.Log.Infof("zfs: DeleteSnapshot called %s", snapshotID)
+	if snapshotID == "" {
+		p.Log.Warning("zfs: Empty snapshotID")
+		return nil
+	}
+
+	return p.deleteBackup(snapshotID)
+}
+
+// GetVolumeID returns the specific identifier for the PersistentVolume.
+func (p *Plugin) GetVolumeID(unstructuredPV runtime.Unstructured) (string, error) {
```

Comment:
need to check if PV is provisioned by zfs-localpv driver or not. Refer: https://github.com/openebs/velero-plugin/blob/master/pkg/cstor/cstor.go#L275